### PR TITLE
Convert H2 driver to honey-sql-2

### LINKED ISSUE #28157: Migrate `:h2` driver to Honey SQL 2

### PR DESCRIPTION
**Pull request by snoe**
Part of https://github.com/metabase/metabase/issues/28009

Resolves [#28157](https://github.com/metabase/metabase/issues/28157)

I don't think there's anything special here, added a test against a `:sql` driver which might be illegal but I figured it's the best way to capture the intent of checking the compilation of a v1 and v2 query.

### CODE REVIEW

**Comment by camsaul** [comment]
> added a test against a `:sql` driver which might be illegal

we have lots of tests against `:sql` in stuff like `driver.sql.query-processor-test`. So not illegal

**Review by camsaul — APPROVED** [review]

File: src/metabase/util/honey_sql_2.clj [hunk 1]
```diff
@@ -32,7 +32,7 @@
   (sqlize [this]
     (let [numerator   (.numerator ^clojure.lang.Ratio this)
           denominator (.denominator ^clojure.lang.Ratio this)]
-      (clojure.core/format "(%d / %d)" numerator denominator))))
+      (clojure.core/format "(%d.0 / %d.0)" numerator denominator))))
```
Comment (snoe) [inline src/metabase/util/honey_sql_2.clj:35]: `select 1/2` in h2 returns 0 so we get divide by zero errors.

File: src/metabase/driver/sql/query_processor.clj [hunk 2]
```diff
@@ -747,7 +747,9 @@
   [driver [_ arg pred]]
   (hx/call :sum (hx/call :case
                     (->honeysql driver pred) (->honeysql driver arg)
-                    :else                    0.0)))
+                    :else                    (case (long hx/*honey-sql-version*)
+                                               1 0.0 field-alias
```
Comment (camsaul) [inline src/metabase/driver/sql/query_processor.clj:751]: this line looks wrong

File: src/metabase/util/honey_sql_2.clj (hunk 1 above)
Comment (camsaul) [reply src/metabase/util/honey_sql_2.clj:35]: that makes sense, maybe we should just change the inline behavior for Ratio to call `double` on it

**Review by camsaul — APPROVED** [review]

**Pull request merged**